llvm/lib/IR/Metadata.cpp

```diff
@@ -152,26 +152,47 @@ DPValue *DebugValueUser::getUser() { return static_cast<DPValue *>(this); }
 const DPValue *DebugValueUser::getUser() const {
   return static_cast<const DPValue *>(this);
 }
-void DebugValueUser::handleChangedValue(Metadata *NewMD) {
-  getUser()->handleChangedLocation(NewMD);
+
+void DebugValueUser::handleChangedValue(void *Old, Metadata *New) {
+  // NOTE: We could inform the "owner" that a value has changed through
+  // getOwner, if needed.
+  auto OldMD = static_cast<Metadata **>(Old);
+  ptrdiff_t Idx = std::distance(DebugValues.begin(), OldMD);
+  resetDebugValue(Idx, New);
```

Comment:
What's the probability that we could have duplicate metadatas, i.e. multiple entries here? I think it's idiomatic for list-heads to assign themselves with their own address when they're empty, would that show up here as same-value same-address Metadatas?

Reply:
Do you mean duplicate metadata as-in, the same metadata appearing multiple times in this list? If so, the answer is that we ignore that here - this function should be called once for each pointer to the metadata being replaced that's being tracked anywhere, so if the pointer appears twice in here then they'll each have their own call even if the `Owner` is the same.

Reply:
Hmmm, are there any alternatives to this? The method name and signature feel mismatched with what it does, in that calling this function will replace _a_ Old/New metadata pair, but possibly not the one that the caller expected.

To be clear, the code is fine, it's the risk of someone reading this and either being confused or having an incorrect understanding that I worry about.

Reply:
`handleChangedValuePointer`? In any case, the metadata RAUW mechanism is based on maintaining a map of addresses of pointers to that metadata, so if two pointers to the same MD exist in DebugValueUser, we would expect this to be called twice. We could iterate over the full user list and reset every identical pointer here, but `ReplaceableMetadataImpl::replaceAllUsesWith()` (which is the only thing that should call through to this) already has reference to all of those pointers, so I think this would just be duplicating work.

Reply:
Cool; IMO best to document this in a comment then, that it's expected to operate in this way.